py/objint_mpz.c

```diff
@@ -107,15 +107,77 @@ char *mp_obj_int_formatted_impl(char **buf, size_t *buf_size, size_t *fmt_size,
     return str;
 }
 
-mp_obj_t mp_obj_int_from_bytes_impl(bool big_endian, size_t len, const byte *buf) {
+mp_obj_t mp_obj_int_from_bytes_impl(bool big_endian, bool arg_signed, size_t len, const byte *buf) {
     mp_obj_int_t *o = mp_obj_int_new_mpz();
     mpz_set_from_bytes(&o->mpz, big_endian, len, buf);
+    if (arg_signed) {
+        size_t msb = big_endian ? 0 : len-1;
+        if (buf[msb] & 0x80) {
+            o->mpz.neg = 1;
+        }
+    }
```

Comment:
I think this negation logic belongs in `mpz_set_from_bytes()`, it should be slightly smaller code there.